docs/design/storage-auto-scale.md

```diff
@@ -0,0 +1,120 @@
+
+2) If the expansion is not in progress set the status `phase` to `NotStarted` 
+   
+3) If an expansion is in progress(expectedOsdSize!=startOsdSize || expectedOsdCount!=startOsdCount), check the progress and then requeue each 1 minute until the expansion is completed successfully(jump to step 9).
+   
+4) If the highest osd percentage reported in the sync map is more than osdScalingThresholdPercent(70%) means reaching osd nearfull, scaling is needed.
+   
+5) Check if the totalCapacity allows to scale it or not.
+        
+    1) If yes, Update `phase` to `InProgress`  and `lastExpansionStartTime` as `current-time` on the `StorageAutoScaling` CR which need scaling.
+        
+    2) If no, don't reconcile further and change the `phase` to `Failed`
+
+6) If scaling is needed calculate the `expectedOsdSize` and `expectedOsdCount`.
+       
+    1)  If the Osd size is less than maxOsdSize(default:8Tib), do vertical scaling by doubling the each osd sizes for that device class.
+            
+    2)  If the Osd size is equal to maxOsdSize(default:8Tib), do a horizontal scaling, by adding 1 osd of maxOsdSize(default:8Tib) on each `storageDeviceSet`.
+   
+7) Update the status, set the `expectedOsdSize` and `expectedOsdCount` to reflect the new expected value and also set `startOsdSize` and `startOsdCount` with current storagecluster values.
+   
+8) Scale by patching the Storagecluster, with all the device sets update needed at the same time.
+
+9)  Verify and Alert:
+   
+   1) Verify the Storagecluster whether the new osds are added or scaled in size, for all the device sets.
+ 
+   2) If the scaling is successful will raise a `Succeeded` alert and update the status of the `StorageAutoScaling` CR with `lastExpansionCompletionTime` and `phase` and also osd count and size.
+   
+   3) If it is fails will do a requeue every 1 min and, will raise a `failure` alert and change the phase to `failed` if scaling not `Succeeded` with in timeoutSeconds(default:1800) interval.
+
+### InProgress state:
+
+Based on the above algorithm there would be two conditions where in-progress is set, elaborating those conditions,
+
+1) If scaling is just started:
+    1) Set `phase` to `InProgress`.
+    2) Verify is the scaling is successful. 
+    3) If the scaling is successful set the `phase` to `completed`.
+    4) If the scaling is un-successful requeue every 1 min, we have the 2nd case. 
+
+2) If the scaling has already started and its requeue
+    1) Now the requeue will happen every 1 min.
+    2) At the start of reconcile will match that `startOsdSize` and `expectedOsdSize` is not equal and similar for osd count.
+    3) And another validation will do is equating storagecluster spec with prometheus response.
+    4) Will requeue till the scaling is in-progress.
+    5) If the scaling is in-progress with more than timeoutSeconds(default:1800) interval we will raise a user `failure` alert.
```

Comment:
Separately, let's be clear about how alerts are raised. OCS operator doesn't do that, right?
```suggestion
    5) If the scaling is in-progress with more than timeoutSeconds(default:1800) interval we set the phase to `failed`.
```